src/compute/src/row_spine.rs

```diff
@@ -622,3 +627,893 @@ pub(crate) fn offset_list_size(data: &OffsetList, mut callback: impl FnMut(usize
+
+        /// Independently encodes each column.
+        #[derive(Default, Debug)]
+        pub struct ColumnsCodec {
+            columns: Vec<DictionaryCodec>,
+            bytes: usize,
+            total: usize,
+        }
+
+        impl Codec for ColumnsCodec {
+            type DecodeIter<'a> = ColumnsIter<'a>;
+            fn decode<'a>(&'a self, bytes: &'a [u8]) -> Self::DecodeIter<'a> {
+                ColumnsIter {
+                    index: Some(self),
+                    column: 0,
+                    data: bytes,
+                }
+            }
+            fn encode<'a, I>(&mut self, iter: I, output: &mut Vec<u8>)
+            where
+                I: IntoIterator<Item = &'a [u8]>,
+            {
+                let mut iter = iter.into_iter();
+                let mut index = 0;
+                while let Some(bytes) = iter.next() {
+                    self.total += bytes.len();
+                    if self.columns.len() <= index {
+                        self.columns.push(Default::default());
+                    }
+                    self.columns[index].encode(std::iter::once(bytes), output);
+                    index += 1;
+                }
+                self.bytes += output.len();
+            }
+
+            fn new_from<'a>(stats: impl IntoIterator<Item = &'a Self>) -> Self {
+                // Is it possible that one of the inputs has no stats?
+                let stats = stats.into_iter().collect::<Vec<_>>();
+                let cols = stats.iter().map(|s| s.columns.len()).max().unwrap_or(0);
+                let mut columns = Vec::with_capacity(cols);
+                let default: DictionaryCodec = Default::default();
+                for index in 0..cols {
+                    columns.push(DictionaryCodec::new_from(
+                        stats
+                            .iter()
+                            .map(|s| s.columns.get(index).unwrap_or(&default)),
+                    ));
+                }
+                Self {
+                    columns,
+                    total: 0,
+                    bytes: 0,
+                }
+            }
+            fn report(&self) {
+                if self.total > 500000 {
+                    //} && self.columns.iter().all(|c| c.decode.len() > 0) {
+                    println!(
+                        "REPORT: {:?} -> {:?} (x{:?})",
+                        self.total,
+                        self.bytes,
+                        self.total / self.bytes
+                    );
+                    println!("COLUMNS: {:?}", self.columns.len());
+                    for column in self.columns.iter() {
+                        column.report()
+                    }
+                }
+            }
+
+            fn borrow_row(row: &Row) -> Self::DecodeIter<'_> {
+                ColumnsIter {
+                    index: None,
+                    column: 0,
+                    data: row.data(),
+                }
```

Comment:
This could call `without_codec` instead. Would require `unsafe` and the associated safety argument, but then we'd have a comment as to why this is safe.